content/asm_1.md

```diff
@@ -1,125 +1,281 @@
+    ;; Reference to the entry point of our program
     global _start
+
+;; Entry point
 _start:
+    ;; Specify the number of the system call (1 is `sys_write`).
     mov     rax, 1
+    ;; Set the first argument of `sys_write` to 1 (`stdout`).
     mov     rdi, 1
+    ;; Set the second argument of `sys_write` to the reference of the `msg` variable.
     mov     rsi, msg
+    ;; Set the third argument to the length of the `msg` variable's value (13 bytes).
     mov     rdx, 13
+
+    ;; Call the `sys_write` system call.
     syscall
+    ;; Specify the number of the system call (60 is `sys_exit`).
     mov    rax, 60
+    ;; Set the first argument of `sys_exit` to `0`. The `0` status code is success.
     mov    rdi, 0
+    ;; Call the `sys_exit` system call.
     syscall
 ```
 
-Yes, it doesn't look like printf("Hello world"). Let's try to understand what is it and how it works. Take a look 1-2 lines. We defined data section and put there msg constant with Hello world value. Now we can use this constant in our code. Next is declaration text section and entry point of program. Program will start to execute from 7 line. Now starts the most interesting part. We already know what is it mov instruction, it gets 2 operands and put value of second to first. But what is it these rax, rdi and etc... As we can read in the wikipedia:
+This may look quite long compared to the "Hello, World!" program written using a high-level programming language. Let’s break it down and understand how it works.
+
+### Defining sections
+
+Have a look at the first four lines of the program. At the beginning we defined the `data` section and specified the `msg` variable with the `hello, world!` value. Now we can use it in the program's code.
+
+Next is the declaration of the `text` section and the `_start` entry point of the program. After running the program, it starts from the `_start` line.
+
+### CPU registers and system calls
+
+After defining the program's sections, we can move to the actual code of the program. The first four lines after section definitions start from the `mov` instruction to place specific values into registers. This instruction expects two operands and puts the value of the second operand in the first one. But what are these `rax`, `rdi`, and `rsi`? We can read in the Wikipedia:
+
+> A central processing unit (CPU) is the hardware within a computer that carries out the instructions of a computer program by performing the basic arithmetical, logical, and input/output operations of the system.
+
+A CPU executes different operations on data. But where does it store data? While the [main memory](https://en.wikipedia.org/wiki/Computer_memory) is an obvious choice, accessing it is relatively slow. Reading and storing data in main memory slows down the operations because it involves complicated steps to send data requests through the control bus. To speed things up, the CPU uses small, fast storage locations called **general-purpose registers**.
```

Comment:
```suggestion
A CPU executes different operations on data. But where is data stored? While the [main memory](https://en.wikipedia.org/wiki/Computer_memory) is an obvious choice, accessing it is relatively slow. Reading and storing data in the main memory slows down the operations because it involves complicated steps to send data requests through the control bus. To speed things up, the CPU uses small, fast storage locations called **general-purpose registers**.
```